README.md
edx_irc_chat
============

IRC server/web client integrated with edX CS169.2x class in August 2013.


ircd-hybrid/ChangeLog
2006-07-17 02:30:03  lusky

	* [r726] include/patchlevel.h:
	  bump version to 7.2.2

2006-07-16 15:41:33  adx

	* [r723] src/hostmask.c:
	  + fixed a bug where auths/klines/glines with digit wildcards (#)
	    were not handled correctly

2006-07-15 12:23:07  michael

	* [r718] messages/Makefile.in:
	  - Updated messages/Makefile.in for ircd-brazilian_pt.lang

2006-07-15 06:22:58  adx

	* [r716] src/s_user.c:
	  + added DEAF isupport

2006-07-14 17:01:55  michael

	* [r715] Makefile.in, configure, configure.ac, contrib/Makefile.in, 
	  contrib/help/Makefile.in, doc/Makefile.in, etc/Makefile.in, 
	  help/Makefile.in, include/setup.h.in, lib/Makefile.in, 
	  lib/pcre/Makefile.in, messages/Makefile.in, 
	  modules/Makefile.in, servlink/Makefile.in, src/Makefile.in, 
	  tools/Makefile.in:
	  - Rebuilt configure using autoconf 2.60

2006-07-14 08:01:10  michael

	* [r714] RELNOTES:
	  - Updated RELNOTES

2006-07-12 08:37:35  michael

	* [r713] src/hook.c:
	  - fixed core on MODRESTART as reported by xZ/CoolCold

2006-07-08 08:02:15  michael

	* [r711] src/ircd_lexer.l, src/lex.yy.c:
	  - Also show the full path of .included configuration files when reporting
	    syntax errors.  This is just cosmetical.

2006-07-08 06:28:31  michael

	* [r710] RELNOTES:
	  - Updated RELNOTES

2006-07-07 04:56:45  michael

	* [r708] src/irc_reslib.c:
	  - Applied nenolod's parse_resvconf() fix

2006-07-02 07:36:17  michael

	* [r707] RELNOTES:
	  - Updated RELNOTES

2006-07-02 07:25:55  michael

	* [r706] modules/core/m_nick.c, src/client.c:
	  - ACCEPT fixes as reported by ThaPrince

2006-06-29 13:05:34  db

	* [r705] messages/ircd-brazilian_pt.lang:
	  - Add Brazilian to 7.2 as well
	    Marcelo Lipienski <und3rl1n3d@gmail.com>

2006-06-22 05:34:55  michael

	* [r698] contrib/ip_cloaking.c:
	  - Fixed core in make_virthost() with host names that doesn't have a dot
	    in it.  Reported by CoolCold

2006-06-21 03:52:16  michael

	* [r696] src/s_auth.c:
	  - Fixed small memory leak in delete_auth()

2006-06-17 07:34:25  michael

	* [r690] contrib/m_force.c:
	  - Backported r689 (the name of the source was appended to the RPL_TOPICTIME
	    numeric instead of the target's name in mo_forcejoin)

2006-06-17 05:46:45  michael

	* [r687] RELNOTES:
	  - Updated RELNOTES

2006-06-16 08:22:31  michael

	* [r685] INSTALL, configure, configure.ac, doxygen.conf, src/version.c:
	  - Updated copyright years in version.c and INSTALL
	  - Bumped version to 7.2.2 in configure.ac and doxygen.conf

2006-06-15 10:28:59  michael

	* [r683] RELNOTES, src/s_bsd.c:
	  - close_connection(): Fixed some harmless bug that could result in incorrect
	    sent statistics in "STATS t"
	  - Updated RELNOTES

2006-06-13 12:51:10  adx

	* [r675] src/channel.c:
	  + patch against truncated NAMES output
	  
	  Submitted by: Murat Deligonul <deligonul@gmail.com>

2006-06-13 03:38:50  michael

	* [r673] src/s_conf.c:
	  - check_classes(): Only call destroy_cidr_class() on a class that is ready
	    for deletion, otherwise the current cidr count would become unadjusted in
	    case a disabled class gets re-enabled later on.

2006-06-13 03:28:30  michael

	* [r672] src/s_conf.c:
	  - Fixed possible core introduced by previous commit
	  - Put some asserts into detach_conf()

2006-06-13 03:15:38  michael

	* [r671] include/s_conf.h, src/ircd_parser.y, src/messages.tab, 
	  src/s_conf.c, src/y.tab.c:
	  - Fixed class limit bug as spotted by stu.  ("REHASH" would always reset
	    current user count)
	  - Made "STATS y|Y" show inactive (deleted) classes that still have attached
	    clients

2006-06-12 08:20:55  michael

	* [r670] RELNOTES, include/balloc.h, src/balloc.c:
	  - balloc.(c|h): backported r544 (Killed Block::used_list)
	  - Update RELNOTES

2006-06-11 09:39:25  michael

	* [r668] lib/pcre/README, src/y.tab.c, src/y.tab.h:
	  - Rebuilt y.tab.(c|h) using bison 2.3
	  - Bumped pcrelib version to 6.4.  There were no changes since 6.3 that would
	    have affected our hacked version of the pcre library, so this is mainly to
	    satisfy paranoid I-always-need-the-latest-version people

2006-06-10 19:35:38  stu

	* [r666] modules/core/m_message.c:
	  Fix compile error.  Be more careful with the so called stable branch :)

2006-06-10 10:34:01  michael

	* [r664] src/s_auth.c:
	  - Fixed compile error in s_auth.c

2006-06-09 14:51:22  adx

	* [r660] include/numeric.h, messages/ircd-dutch.lang, 
	  messages/ircd-german.lang, messages/ircd-polish.lang, 
	  messages/ircd-standard.example.lang, 
	  messages/ircd-swedish.lang, modules/m_whois.c, 
	  src/messages.tab:
	  + backported numeric changes

2006-06-08 03:00:17  michael

	* [r650] include/client.h, src/client.c, src/s_auth.c:
	  - Fixed bug that would exit unknown clients in check_unknowns_list() that are
	    still processing auth/dns queries.
	    Basically check_unknowns_list() should only exit clients that have finished
	    auth/dns and didn't send a NICK/USER combination since they have been
	    released from auth.
	  
	    Reported by ThaPrince.

2006-06-07 08:42:00  michael

	* [r646] modules/m_set.c:
	  - Fixed old bug that would forbid SPAMNUM to be disabled via "SET SPAMNUM 0"

2006-06-01 10:12:21  michael

	* [r639] etc/example.conf, etc/example.efnet.conf:

ircd-hybrid/src/ircd.c
/*
 *  ircd-hybrid: an advanced Internet Relay Chat Daemon(ircd).
 *  ircd.c: Starts up and runs the ircd.
 *
 *  Copyright (C) 2002 by the past and present ircd coders, and others.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 *  USA
 *
 *  $Id: ircd.c 163 2005-10-20 21:09:02Z adx $
 */

#include "stdinc.h"
#include "s_user.h"
#include "tools.h"
#include "ircd.h"
#include "channel.h"
#include "channel_mode.h"
#include "client.h"
#include "common.h"
#include "event.h"
#include "fdlist.h"
#include "hash.h"
#include "irc_string.h"
#include "sprintf_irc.h"
#include "ircd_signal.h"
#include "list.h"
#include "s_gline.h"
#include "motd.h"
#include "ircd_handler.h"
#include "msg.h"         /* msgtab */
#include "hostmask.h"
#include "numeric.h"
#include "packet.h"
#include "parse.h"
#include "irc_res.h"
#include "restart.h"
#include "s_auth.h"
#include "s_bsd.h"
#include "s_conf.h"
#include "s_log.h"
#include "s_misc.h"
#include "s_serv.h"      /* try_connections */
#include "s_stats.h"
#include "send.h"
#include "whowas.h"
#include "modules.h"
#include "memory.h"
#include "hook.h"
#include "ircd_getopt.h"
#include "balloc.h"
#include "motd.h"
#include "supported.h"

/* Try and find the correct name to use with getrlimit() for setting the max.
 * number of files allowed to be open by this process.
 */

/* /quote set variables */
struct SetOptions GlobalSetOptions;

/* configuration set from ircd.conf */
struct config_file_entry ConfigFileEntry; 
/* server info set from ircd.conf */
struct server_info ServerInfo;
/* admin info set from ircd.conf */
struct admin_info AdminInfo = { NULL, NULL, NULL };
struct Counter Count = { 0, 0, 0, 0, 0, 0, 0, 0 };
struct ServerState_t server_state = { 0 };
struct logging_entry ConfigLoggingEntry = { 1, {0}, {0}, {0}, {0}, {0}, {0}, {0}, {0} }; 
struct timeval SystemTime;
struct Client me;             /* That's me */
struct LocalUser meLocalUser; /* That's also part of me */
unsigned long connect_id = 0; /* unique connect ID */

static unsigned long initialVMTop = 0;   /* top of virtual memory at init */
const char *logFileName = LPATH;
const char *pidFileName = PPATH;
const char *activityLogFileName = APATH;

char **myargv;
char ircd_platform[PLATFORMLEN];

int dorehash = 0;
int doremotd = 0;
time_t nextconnect = 1;       /* time for next try_connections call */

/* Set to zero because it should be initialized later using
 * initialize_server_capabs
 */
int default_server_capabs = 0;

#ifdef HAVE_LIBCRYPTO
int bio_spare_fd = -1;
#endif

int splitmode;
int splitchecking;
int split_users;
unsigned int split_servers;

/* Do klines the same way hybrid-6 did them, i.e. at the
 * top of the next io_loop instead of in the same loop as
 * the klines are being applied.
 *
 * This should fix strange CPU starvation as very indirectly reported.
 * (Why do you people not email bug reports? WHY? WHY?)
 *
 * - Dianora
 */

int rehashed_klines = 0;

/*
 * get_vm_top - get the operating systems notion of the resident set size
 */
#ifndef _WIN32
static unsigned long
get_vm_top(void)
{
  /*
   * NOTE: sbrk is not part of the ANSI C library or the POSIX.1 standard
   * however it seems that everyone defines it. Calling sbrk with a 0
   * argument will return a pointer to the top of the process virtual
   * memory without changing the process size, so this call should be
   * reasonably safe (sbrk returns the new value for the top of memory).
   * This code relies on the notion that the address returned will be an 
   * offset from 0 (NULL), so the result of sbrk is cast to a size_t and 
   * returned. We really shouldn't be using it here but...
   */

  void *vptr = sbrk(0);
  return((unsigned long)vptr);
}

/*
 * print_startup - print startup information
 */
static void
print_startup(int pid)
{
  printf("ircd: version %s\n", ircd_version);
  printf("ircd: pid %d\n", pid);
  printf("ircd: running in %s mode from %s\n", !server_state.foreground ? "background"
         : "foreground", ConfigFileEntry.dpath);
}

static void
make_daemon(void)
{
  int pid;

  if ((pid = fork()) < 0)
  {
    perror("fork");
    exit(EXIT_FAILURE);
  }
  else if (pid > 0)
  {
    print_startup(pid);
    exit(EXIT_SUCCESS);
  }

  setsid();
}
#endif

/*
 * get_maxrss - get the operating systems notion of the resident set size
 */
unsigned long
get_maxrss(void)
{
#ifdef _WIN32
  return (0);   /* FIXME */
#else
  return (get_vm_top() - initialVMTop);
#endif
}

static int printVersion = 0;

struct lgetopt myopts[] = {
  {"dlinefile",  &ConfigFileEntry.dlinefile, 
   STRING, "File to use for dline.conf"},
  {"configfile", &ConfigFileEntry.configfile, 
   STRING, "File to use for ircd.conf"},
  {"klinefile",  &ConfigFileEntry.klinefile, 
   STRING, "File to use for kline.conf"},
  {"xlinefile",  &ConfigFileEntry.xlinefile, 
   STRING, "File to use for xline.conf"},
  {"logfile",    &logFileName, 
   STRING, "File to use for ircd.log"},
  {"activitylogfile", &activityLogFileName,
   STRING, "File to use for logging all activity"},
  {"pidfile",    &pidFileName,
   STRING, "File to use for process ID"},
  {"foreground", &server_state.foreground, 
   YESNO, "Run in foreground (don't detach)"},
  {"version",    &printVersion, 
   YESNO, "Print version and exit"},
  {"help", NULL, USAGE, "Print this text"},
  {NULL, NULL, STRING, NULL},
};

void
set_time(void)
{
  static char to_send[200];
  struct timeval newtime;
#ifdef _WIN32
  FILETIME ft;

  GetSystemTimeAsFileTime(&ft);
  if (ft.dwLowDateTime < 0xd53e8000)
    ft.dwHighDateTime--;
  ft.dwLowDateTime -= 0xd53e8000;
  ft.dwHighDateTime -= 0x19db1de;

  newtime.tv_sec  = (*(uint64_t *) &ft) / 10000000;
  newtime.tv_usec = (*(uint64_t *) &ft) / 10 % 1000000;
#else
  newtime.tv_sec  = 0;
  newtime.tv_usec = 0;

  if (gettimeofday(&newtime, NULL) == -1)
  {
    ilog(L_ERROR, "Clock Failure (%s), TS can be corrupted",
         strerror(errno));
    sendto_realops_flags(UMODE_ALL, L_ALL,
                         "Clock Failure (%s), TS can be corrupted",
                         strerror(errno));
    restart("Clock Failure");
  }
#endif

  if (newtime.tv_sec < CurrentTime)
  {
    ircsprintf(to_send, "System clock is running backwards - (%lu < %lu)",
               (unsigned long)newtime.tv_sec, (unsigned long)CurrentTime);
    report_error(L_ALL, to_send, me.name, 0);
    set_back_events(CurrentTime - newtime.tv_sec);
  }

  SystemTime.tv_sec  = newtime.tv_sec;
  SystemTime.tv_usec = newtime.tv_usec;
}

static void
io_loop(void)
{
  while (1 == 1)
  {
    /*
     * Maybe we want a flags word?
     * ie. if (REHASHED_KLINES(global_flags)) 
     * SET_REHASHED_KLINES(global_flags)
     * CLEAR_REHASHED_KLINES(global_flags)
     *
     * - Dianora
     */
    if (rehashed_klines)
    {
      check_conf_klines();
      rehashed_klines = 0;
    }

    if (listing_client_list.head)
    {
      dlink_node *ptr = NULL, *ptr_next = NULL;
      DLINK_FOREACH_SAFE(ptr, ptr_next, listing_client_list.head)
      {
        struct Client *client_p = ptr->data;
        assert(client_p->localClient->list_task);
        safe_list_channels(client_p, client_p->localClient->list_task, 0, 0);
      }
    }

    /* Run pending events, then get the number of seconds to the next
     * event
     */
    while (eventNextTime() <= CurrentTime)
      eventRun();

    comm_select();
    exit_aborted_clients();
    free_exited_clients();
    send_queued_all();

    /* Check to see whether we have to rehash the configuration .. */
    if (dorehash)
    {
      rehash(1);
      dorehash = 0;
    }
    if (doremotd)
    {
      read_message_file(&ConfigFileEntry.motd);
      sendto_realops_flags(UMODE_ALL, L_ALL,
                           "Got signal SIGUSR1, reloading ircd motd file");
      doremotd = 0;
    }
  }
}

/* initalialize_global_set_options()
 *
 * inputs       - none
 * output       - none
 * side effects - This sets all global set options needed 
 */
static void
initialize_global_set_options(void)
{
  memset(&GlobalSetOptions, 0, sizeof(GlobalSetOptions));

  GlobalSetOptions.autoconn  = 1;
  GlobalSetOptions.spam_time = MIN_JOIN_LEAVE_TIME;
  GlobalSetOptions.spam_num  = MAX_JOIN_LEAVE_COUNT;

  if (ConfigFileEntry.default_floodcount)
    GlobalSetOptions.floodcount = ConfigFileEntry.default_floodcount;
  else
    GlobalSetOptions.floodcount = 10;

  /* XXX I have no idea what to try here - Dianora */
  GlobalSetOptions.joinfloodcount = 16;
  GlobalSetOptions.joinfloodtime = 8;

  split_servers = ConfigChannel.default_split_server_count;
  split_users   = ConfigChannel.default_split_user_count;

  if (split_users && split_servers && (ConfigChannel.no_create_on_split ||
                                       ConfigChannel.no_join_on_split))
  {
    splitmode     = 1;
    splitchecking = 1;
  }

  GlobalSetOptions.ident_timeout = IDENT_TIMEOUT;
  GlobalSetOptions.idletime = ConfigFileEntry.idletime;
  /* End of global set options */
}

/* initialize_message_files()
 *
 * inputs       - none
 * output       - none
 * side effects - Set up all message files needed, motd etc.
 */
static void
initialize_message_files(void)
{
  init_message_file(USER_MOTD, MPATH, &ConfigFileEntry.motd);
  init_message_file(OPER_MOTD, OPATH, &ConfigFileEntry.opermotd);
  init_message_file(USER_LINKS, LIPATH, &ConfigFileEntry.linksfile);

  read_message_file(&ConfigFileEntry.motd);
  read_message_file(&ConfigFileEntry.opermotd);
  read_message_file(&ConfigFileEntry.linksfile);

  init_isupport();
}

/* initialize_server_capabs()
 *
 * inputs       - none
 * output       - none
 */
static void
initialize_server_capabs(void)
{
  add_capability("QS", CAP_QS, 1);
  add_capability("LL", CAP_LL, 1);
  add_capability("EOB", CAP_EOB, 1);
  if (ServerInfo.sid != NULL)	/* only enable TS6 if we have an SID */
    add_capability("TS6", CAP_TS6, 0);
  add_capability("ZIP", CAP_ZIP, 0);
  add_capability("CLUSTER", CAP_CLUSTER, 1);
#ifdef HALFOPS
  add_capability("HOPS", CAP_HOPS, 1);
#endif
}

/* write_pidfile()
 *
 * inputs       - filename+path of pid file
 * output       - NONE
 * side effects - write the pid of the ircd to filename
 */
static void
write_pidfile(const char *filename)
{
  FBFILE *fb;

  if ((fb = fbopen(filename, "w")))
  {
    char buff[32];
    unsigned int pid = (unsigned int)getpid();
    size_t nbytes = ircsprintf(buff, "%u\n", pid);

    if ((fbputs(buff, fb, nbytes) == -1))
      ilog(L_ERROR, "Error writing %u to pid file %s (%s)",
           pid, filename, strerror(errno));

    fbclose(fb);
    return;
  }
  else
  {
    ilog(L_ERROR, "Error opening pid file %s", filename);
  }
}

/* check_pidfile()
 *
 * inputs       - filename+path of pid file
 * output       - none
 * side effects - reads pid from pidfile and checks if ircd is in process
 *                list. if it is, gracefully exits
 * -kre
 */
static void
check_pidfile(const char *filename)
{
#ifndef _WIN32
  FBFILE *fb;
  char buff[32];
  pid_t pidfromfile;

  /* Don't do logging here, since we don't have log() initialised */
  if ((fb = fbopen(filename, "r")))
  {
    if (fbgets(buff, 20, fb) == NULL)
    {
      /* log(L_ERROR, "Error reading from pid file %s (%s)", filename,
       * strerror(errno));
       */
    }
    else
    {
      pidfromfile = atoi(buff);

      if (!kill(pidfromfile, 0))
      {
        /* log(L_ERROR, "Server is already running"); */
        printf("ircd: daemon is already running\n");
        exit(-1);
      }
    }

    fbclose(fb);
  }
  else if (errno != ENOENT)
  {
    /* log(L_ERROR, "Error opening pid file %s", filename); */
  }
#endif
}

/* setup_corefile()
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups corefile to system limits.
 * -kre
 */
static void
setup_corefile(void)
{
#ifdef HAVE_SYS_RESOURCE_H
  struct rlimit rlim; /* resource limits */

  /* Set corefilesize to maximum */
  if (!getrlimit(RLIMIT_CORE, &rlim))
  {
    rlim.rlim_cur = rlim.rlim_max;
    setrlimit(RLIMIT_CORE, &rlim);
  }
#endif
}

/* init_ssl()
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups SSL context.
 */
static void
init_ssl(void)
{
#ifdef HAVE_LIBCRYPTO
  SSL_load_error_strings();
  SSLeay_add_ssl_algorithms();

  ServerInfo.ctx = SSL_CTX_new(SSLv23_server_method());
  if (!ServerInfo.ctx)
  {
    const char *s;

    fprintf(stderr, "ERROR: Could not initialize the SSL context -- %s\n",
            s = ERR_lib_error_string(ERR_get_error()));
    ilog(L_CRIT, "ERROR: Could not initialize the SSL context -- %s\n", s);
  }

  SSL_CTX_set_options(ServerInfo.ctx, SSL_OP_NO_SSLv2);
  SSL_CTX_set_options(ServerInfo.ctx, SSL_OP_TLS_ROLLBACK_BUG|SSL_OP_ALL);
  SSL_CTX_set_verify(ServerInfo.ctx, SSL_VERIFY_NONE, NULL);

  bio_spare_fd = save_spare_fd("SSL private key validation");
#endif /* HAVE_LIBCRYPTO */
}

/* init_callbacks()
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups standard hook points
 */
static void
init_callbacks(void)
{
  iorecv_cb = register_callback("iorecv", iorecv_default);
  iosend_cb = register_callback("iosend", iosend_default);
  iorecvctrl_cb = register_callback("iorecvctrl", NULL);
  iosendctrl_cb = register_callback("iosendctrl", NULL);
}

int
main(int argc, char *argv[])
{
  /* Check to see if the user is running
   * us as root, which is a nono
   */
#ifndef _WIN32
  if (geteuid() == 0)
  {
    fprintf(stderr, "Don't run ircd as root!!!\n");
    return(-1);
  }

  /* Setup corefile size immediately after boot -kre */
  setup_corefile();

  /* set initialVMTop before we allocate any memory */
  initialVMTop = get_vm_top();
#endif

  /* save server boot time right away, so getrusage works correctly */
  set_time();

    /* It ain't random, but it ought to be a little harder to guess */
  srand(SystemTime.tv_sec ^ (SystemTime.tv_usec | (getpid() << 20)));
  memset(&me, 0, sizeof(me));
  memset(&meLocalUser, 0, sizeof(meLocalUser));
  me.localClient = &meLocalUser;
  dlinkAdd(&me, &me.node, &global_client_list);	/* Pointer to beginning
						   of Client list */

  memset(&ServerInfo, 0, sizeof(ServerInfo));

  /* Initialise the channel capability usage counts... */
  init_chcap_usage_counts();

  ConfigFileEntry.dpath      = DPATH;
  ConfigFileEntry.configfile = CPATH;  /* Server configuration file */
  ConfigFileEntry.klinefile  = KPATH;  /* Server kline file         */
  ConfigFileEntry.xlinefile  = XPATH;  /* Server xline file         */
  ConfigFileEntry.rxlinefile = RXPATH; /* Server regex xline file   */
  ConfigFileEntry.rklinefile = RKPATH; /* Server regex kline file   */
  ConfigFileEntry.dlinefile  = DLPATH; /* dline file                */
  ConfigFileEntry.glinefile  = GPATH;  /* gline log file            */
  ConfigFileEntry.cresvfile  = CRESVPATH; /* channel resv file      */
  ConfigFileEntry.nresvfile  = NRESVPATH; /* nick resv file         */
  myargv = argv;
  umask(077);                /* better safe than sorry --SRB */

  parseargs(&argc, &argv, myopts);

  if (printVersion)
  {
    printf("ircd: version %s\n", ircd_version);
    exit(EXIT_SUCCESS);
  }

  if (chdir(ConfigFileEntry.dpath))
  {
    perror("chdir");
    exit(EXIT_FAILURE);
  }

  init_ssl();

#ifndef _WIN32
  if (!server_state.foreground)
  {
    make_daemon();
    close_standard_fds(); /* this needs to be before init_netio()! */
  }
  else
    print_startup(getpid());

  setup_signals();
#endif

  get_ircd_platform(ircd_platform);

  /* Init the event subsystem */
  eventInit();
  /* We need this to initialise the fd array before anything else */
  fdlist_init();
  init_log(logFileName);
  open_activity_log(activityLogFileName);
  check_can_use_v6();
  init_comm();         /* This needs to be setup early ! -- adrian */
  /* Check if there is pidfile and daemon already running */
  check_pidfile(pidFileName);

#ifndef NOBALLOC
  initBlockHeap();
#endif
  init_dlink_nodes();
  init_callbacks();
  initialize_message_files();
  dbuf_init();
  init_hash();
  init_ip_hash_table();      /* client host ip hash table */
  init_host_hash();          /* Host-hashtable. */
  clear_tree_parse();
  init_client();
  init_class();
  init_whowas();
  init_stats();
  read_conf_files(1);   /* cold start init conf files */
  initServerMask();
  me.id[0] = '\0';
  init_uid();
  init_auth();          /* Initialise the auth code */
#ifndef _WIN32
  init_resolver();      /* Needs to be setup before the io loop */
#endif
  initialize_server_capabs();   /* Set up default_server_capabs */
  initialize_global_set_options();
  init_channels();

  if (ServerInfo.name == NULL)
  {
    ilog(L_CRIT, "No server name specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }
  strlcpy(me.name, ServerInfo.name, sizeof(me.name));

  /* serverinfo{} description must exist.  If not, error out.*/
  if (ServerInfo.description == NULL)
  {
    ilog(L_CRIT,
      "ERROR: No server description specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }
  strlcpy(me.info, ServerInfo.description, sizeof(me.info));

  me.from    = &me;
  me.servptr = &me;

  SetMe(&me);
  make_server(&me);

  me.lasttime = me.since = me.firsttime = CurrentTime;
  hash_add_client(&me);
  
  /* add ourselves to global_serv_list */
  dlinkAdd(&me, make_dlink_node(), &global_serv_list);

  check_class();

#ifndef STATIC_MODULES
  if (chdir(MODPATH))
  {
    ilog (L_CRIT, "Could not load core modules. Terminating!");
    exit(EXIT_FAILURE);
  }

  load_all_modules(1);
  load_conf_modules();
  load_core_modules(1);
  /* Go back to DPATH after checking to see if we can chdir to MODPATH */
  chdir(ConfigFileEntry.dpath);
#else
  load_all_modules(1);
#endif
  /*
   * assemble_umode_buffer() has to be called after
   * reading conf/loading modules.
   */
  assemble_umode_buffer();

  write_pidfile(pidFileName);

  ilog(L_NOTICE, "Server Ready");

  eventAddIsh("cleanup_glines", cleanup_glines, NULL, CLEANUP_GLINES_TIME);
  eventAddIsh("cleanup_tklines", cleanup_tklines, NULL, CLEANUP_TKLINES_TIME);

  /* We want try_connections to be called as soon as possible now! -- adrian */
  /* No, 'cause after a restart it would cause all sorts of nick collides */
  eventAddIsh("try_connections", try_connections, NULL, STARTUP_CONNECTIONS_TIME);

  eventAddIsh("collect_zipstats", collect_zipstats, NULL, ZIPSTATS_TIME);

  /* Setup the timeout check. I'll shift it later :)  -- adrian */
  eventAddIsh("comm_checktimeouts", comm_checktimeouts, NULL, 1);

  if (ConfigServerHide.links_delay > 0)
    eventAddIsh("write_links_file", write_links_file, NULL, ConfigServerHide.links_delay);
  else
    ConfigServerHide.links_disabled = 1;

  if (splitmode)
    eventAddIsh("check_splitmode", check_splitmode, NULL, 60);

  io_loop();
  return(0);
}

/*
 * Activity logging
 *
 * This is a separate logfile for logging all server activity (messages,
 * joins, etc)
 *
 */

static FILE *activity_log_file = NULL;

/* 
 * open_activity_log   - open activity log file
 *
 * inputs - activity log filename
 *
 */

void open_activity_log(const char *filename)
{
  activity_log_file = fopen(filename, "a");

  if (activity_log_file == NULL)
  {
#ifdef USE_SYSLOG
    syslog(LOG_ERR, "Unable to open log file: %s: %s",
           filename, strerror(errno));
#endif
  } else {
    activity_log("START");
  }
}

void activity_log(char *s, ...)
{
  va_list args;

  if (!activity_log_file)
    return;

  va_start(args, s);

  fprintf(activity_log_file, "%i ", time(NULL));
  vfprintf(activity_log_file, s, args);
  fprintf(activity_log_file, "\n");

  fflush(activity_log_file);

  va_end(args);
}

